### ISSUE TITLE
Cache/settings/bundle problem

### ISSUE THREAD

**Issue by dein0s** [zsh-users/antigen#463]
Not sure if this is a bug or just me missing something, so decided to post it here. If this issue doesn't suit the [contributing guide](https://github.com/zsh-users/antigen/blob/develop/CONTRIBUTING.md) - let me know and I'll re-write/close it.
Also, before we continue, it's worth to mention that english aint my native, so I appologize for any misstakes and inconvinience in advance.

**Antigen**: 1.4.1
**ZSH**:  5.1.1

Steps are:
1. Reset antigen with `antigen reset` command
2. Open terminal
3. Check autoseggestions (cmd + tab) - everything is fine
![autosuggestions](http://i.imgur.com/SGzHqCc.png)
4. Close terminal
5. Open terminal
6. Check autosuggestions (cmd + tab) - nothing shows up
![no_autosuggestions](http://i.imgur.com/doGWhD2.png)
7. Reset antigen - everything is fine again

It applies not to all bundles/plugins. But [wd](https://github.com/robbyrussell/oh-my-zsh/tree/master/plugins/wd) is the most popular one for me.

My [.zshrc](https://github.com/dein0s/dotfiles/blob/master/shells/zsh/zshrc.dotfile) (took this part from cookbook w/o detailed understanding):
```bash
export ANTIGEN_COMPDUMPFILE="${HOME}/.zcompdump"

# If there is antigen cache available
if [[ -f "${ADOT:-"${HOME}"/.antigen}/.cache/.zcache-payload" ]]; then
    # Load bundles statically
    source "${ADOT:-"${HOME}"/.antigen}/.cache/.zcache-payload"
    autoload -Uz compinit
    compinit -d "${ANTIGEN_COMPDUMPFILE}"
else
    source "${HOME}/.antigen.zsh"
    antigen init "${HOME}/.antigenrc"
fi
```
My [.antigenrc](https://github.com/dein0s/dotfiles/blob/master/shells/zsh/antigenrc.dotfile):
```bash
antigen use oh-my-zsh
antigen bundles <<EOBUNDLES
    catimg
    chucknorris
    colorize
    command-not-found
    cp
    docker
    extract
    git
    git-extras
    gitfast
    history
    httpie
    per-directory-history
    soimort/translate-shell@develop
    taskwarrior
    thefuck
    web-search
    zsh-users/zsh-autosuggestions
    wd
    sudo
EOBUNDLES
antigen theme dein0s/dotfiles shells/zsh/themes/custom_agnoster
antigen apply
```
Both files and `.antigen.zsh` itself are symlinked into `$HOME` dir and points to appropriate files in my `.dotfiles` folder (antigen cloned as a submodule).

**Comment by desyncr** [comment]
Please verify `"${ADOT:-"${HOME}"` is correct. I believe it should be `"$ADOTDIR`.

**Comment by dein0s** [comment]
In [Cookbook](https://github.com/zsh-users/antigen/wiki/Cookbook#static-bundle-loading) it's specified as `$ADOT`, in [Configuration](https://github.com/zsh-users/antigen#configuration) section - `$ADOTDIR`.
But does this really matter if I have a substitution if nothing is found in current env?

Btw, maybe I need to change the path to the cache file totally? As I see from conf section and my current vars,  cache is currently stored in `$_ANTIGEN_CACHE`, which is differ from `${ADOT:-"${HOME}"/.antigen}/.cache/.zcache-payload`:
```bash
~/.antigen  echo $_ANTIGEN_CACHE 
/home/dein0s/.antigen/init.zsh
```

**Comment by desyncr** [comment]
`init.zsh` is the path for cache from `develop` branch.

the Cookbook section reflects current stable version (`v1.4.1`), and thanks for spotting that error!

Try using the path to the cache file directly rather than using env variables.

**Comment by dein0s** [comment]
> `init.zsh` is the path for cache from `develop` branch.

Yep, I've configured a submodule to watch for `develop` branch, as your `master` is still at 1.3.5. Should've mentioned this before, my bad.

Gonna play with cache file path then and fix submodule at 0bc49f9c9bfd9bbd78e29a2dc1c3f29f9b9a1227. I'll report tomorrow.

**Comment by desyncr** [comment]
@dein0s Master moves even slower than the stable releases. It should be on `v1.4.2` on soon as `develop` is released.

If you're on `develop` then you don't need to use the a custom static loaded as Antigen do it for yourself.

**Comment by dein0s** [comment]
> If you're on develop then you don't need to use a custom static loaded as Antigen do it for yourself.

Can you provide an example of how should my `.zshrc` look like to load Antigen then?

**Comment by desyncr** [comment]
A simple `.zshrc` as the following would do the job:

```
    source /path/to/antigen.zsh
    antigen init /path/to/.antigenrc

```

Do `antigen reset` and `antigen cache-gen`. A file `init.zsh` should be created at `${ADOTDIR:-$HOME/.antigen}`. When you start your shell Antigen will use that file to load bundles and lazyload Antigen itself.

Hope that helps!

**Comment by desyncr** [comment]
Here's the `.zshrc` for `v1.4.1` and below:

```
_ANTIGEN_CACHE_PATH=$ADOTDIR/.cache
ANTIGEN_COMPDUMPFILE="${HOME}/.zcompdump"

# If there is antigen cache available
if [[ -f "$_ANTIGEN_CACHE_PATH/.zcache-payload" ]]; then
    # Load bundles statically
    source "$_ANTIGEN_CACHE_PATH/.zcache-payload"
    autoload -Uz compinit
    compinit -d "$ANTIGEN_COMPDUMPFILE"
else
    source "$HOME/.antigen.zsh"
    antigen init "$HOME/.antigenrc"
fi
```

Now, regarding `ANTIGEN_COMPDUMPFILE`, it may have an issue in `v1.4.1` that is fixed in `develop` (https://github.com/zsh-users/antigen/issues/452).

Let me know if it works for you :)

**Comment by dein0s** [comment]
Got that, I'll stick to the latest `develop` then :) Atleast until the 2.0 release, if you don't mind, as I don't really like to lock submodules on any particular commit, as this functionality is just meh...

One more thing - `antigen reset` and `antigen cache-gen` are one-time-to-run commands, right?

Btw, as an update for the subj - that was definetly some cache problem here. Edited `.zshrc` to source `init.zsh` if it exists, autosuggestions works perfect for the last few hours.

**Comment by desyncr** [comment]
> One more thing - antigen reset and antigen cache-gen are one-time-to-run commands, right?

Yep, those commands are handled by Antigen itself. Only on special cases you should be using them, ie. debugging something.

Antigen does a cache reset (`develop` branch) on config changes if you've configured `_ANTIGEN_CACHE` (See https://github.com/zsh-users/antigen#configuration). `cache-gen` is done by Antigen every time it can't find the cache file.

**Comment by dein0s** [comment]
Okay then, regarding to the last comments:
- cleaned up my `.zshrc`, it now contains only the following to lines that are related to Antigen (from https://github.com/zsh-users/antigen/issues/463#issuecomment-290692515 , as decided that I'll stick with a `develop` branch for now)
```bash
source "${HOME}/.antigen.zsh"
antigen init "${HOME}/.antigenrc"
```
- set symlinks to both of this files
```bash
 ~  ls -la | grep antigen
drwxr-xr-x  3 dein0s dein0s  4096 Mar 31 16:59 .antigen/
lrwxrwxrwx  1 dein0s dein0s    59 Mar 22 00:20 .antigenrc -> /home/dein0s/Personal/dotfiles/shells/zsh/antigenrc.dotfile
lrwxrwxrwx  1 dein0s dein0s    65 Mar 30 05:31 .antigen.zsh -> /home/dein0s/Personal/dotfiles/shells/zsh/antigen/bin/antigen.zsh
```
- submodule is set to track `develop` branch and is up to date
```bash
 ~/…/shells/zsh/antigen   develop  git status
On branch develop
Your branch is up-to-date with 'origin/develop'.
nothing to commit, working tree clean
```
- `rm -rf $HOME/.antigen`
- `rm -rf $HOME/.zcompcache`
- `rm $HOME/.zcompdump`
- `rm $HOME./zcompdump.zwc`
- restart terminal
- bundles installed, autosuggestions working
- restart terminal
- autosuggestions aint working (I guess that cache is not loaded properly)
- cache file exists
```bash
~  ls -la .antigen/
total 56
drwxr-xr-x  3 dein0s dein0s  4096 Mar 31 16:55 ./
drwxr-xr-x 44 dein0s dein0s  4096 Mar 31 16:56 ../
-rw-r--r--  1 dein0s dein0s 12504 Mar 31 16:55 init.zsh
-r--r--r--  1 dein0s dein0s 28528 Mar 31 16:55 init.zsh.zwc
drwxr-xr-x  6 dein0s dein0s  4096 Mar 31 16:55 repos/
```

p.s. I can jump on a gitter channel right now, if you have some spare time for this

**Comment by desyncr** [comment]
@dein0s Jump right in! I have a few mins.

**Comment by dein0s** [comment]
Notes from chat:

> Peter @dein0s 18:21
> allright, after futher investigation i came up with the following:
> `.zcompdump` is not storred correctly for some reason
> `-C` flag for `compinit` does not allow to re-ccreate dump file if there is one already
> 
> not sure, is there any sensitive data stored in .zcompdump file, so i'll create a gist for the diff between working and not working states
> as fas as i can see for now from that diff, two of my plugins missing their autocompletions 

Gist with `init.zsh` and diff for `.zcompdump` https://gist.github.com/dein0s/888230d381542b3bfa517a017fad0091

**Comment by dein0s** [comment]
Looks like this was some local issue with `.zcompdump` file. Redefined it in `.zshrc` and everything works fine now.
```bash
export _ANTIGEN_COMPDUMP="${ADOTDIR:-$HOME/.antigen}/.zcompdump"
```


Also, as I've mentioned in chat - that `-C` flag for `compinit` is worth thinking about.
And we are done here :) Thanks for your support.

Feel free to close this issue if you don't need any additional info from me.

**Comment by desyncr** [comment]
We're currently re-creating zcompdump on apply (ei, when doing `antigen reset`) so `-C` is unnecessary and it won't cause issues with stalled zcompdump files.

**Comment by desyncr** [comment]
Re-opening as it has been reported there may be a regression. https://gitter.im/antigen-zsh/develop?at=5942b726d83c50560c0fbc0e

**Comment by dein0s** [comment]
My config files: [`.zshrc`](https://github.com/dein0s/dotfiles/blob/master/shells/zsh/zshrc.dotfile), [`.antigenrc`](https://github.com/dein0s/dotfiles/blob/master/shells/zsh/antigenrc.dotfile)

**Comment by desyncr** [comment]
Thanks! I'm looking at this :)

**Comment by desyncr** [comment]
@dein0s There is a configuration issue that causes zcompdump to be miss configured. See [here](https://github.com/dein0s/dotfiles/blob/master/shells/zsh/zshrc.dotfile#L3). This was changed in `2.0.0` release, see [changelog](https://github.com/zsh-users/antigen/blob/develop/CHANGELOG.md#200---2017-04-08).

Anyways `antigen-reset` seems to cause this regression, so I'm gonna take a further look.

**Comment by dein0s** [comment]
Yep, noticed that and already changed locally to this:
```bash
# That way Antigen will store it's info in an independent file, so it won't get corrupted by any other script/software
export ADOTDIR="${HOME}/.antigen"
export ANTIGEN_COMPDUMP="${ADOTDIR}/.zcompdump"
export ANTIGEN_LOG="${ADOTDIR}/antigen.log"
export ANTIGEN_DEBUG_LOG="${ADOTDIR}/antigen_debug.log"
# export _ANTIGEN_COMPDUMP="${ADOTDIR:-$HOME/.antigen}/.zcompdump"
# export _ANTIGEN_LOG="${ADOTDIR:-$HOME/.antigen}/antigen.log"
```
Still no success.

**Comment by dein0s** [comment]
Almost the same reason as described in https://github.com/zsh-users/antigen/issues/463#issuecomment-290745592: `-C` for `compinit`

**Comment by desyncr** [comment]
Pushed a fix in #549. It seems to fix the issue described.

Thanks for helping out! @dein0s
